### PR TITLE
[5.0] [Debug] Add parameter type-hints where possible

### PR DESCRIPTION
**Pull request by Matts**
| Q             | A
| ------------- | ---
| Branch?       | 5.0
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no    
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #32179 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

Added scalar typehints to Debug (DebugBundle required no work)

### CODE REVIEW

**Comment by Matts** [comment]
I did not change the BufferingLogger as this extends AbstractLogger -> LoggerInterface, and this is part of the Psr spec. Should this be changed as well? I think this could have big impact?

File: src/Symfony/Component/Debug/ExceptionHandler.php [hunk 1]
```diff
@@ -64,7 +64,7 @@ public function __construct(bool $debug = true, string $charset = null, $fileLin
      *
      * @return static
      */
-    public static function register($debug = true, $charset = null, $fileLinkFormat = null)
+    public static function register(bool $debug = true, string $charset = null, string $fileLinkFormat = null)
```
Comment (fancyweb) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: `$fileLinkFormat` can also be an instance of `FileLinkFormatter`

Comment (Matts) [reply src/Symfony/Component/Debug/ExceptionHandler.php]: Excellent throw. 🏀

File: src/Symfony/Component/Debug/DebugClassLoader.php [hunk 2]
```diff
@@ -128,22 +128,17 @@ public static function disable()
         }
     }
 
-    /**
-     * @return string|null
-     */
-    public function findFile($class)
+    public function findFile(string $class = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: not sure about the change: the annotation looked fine, and no default should be set

File: src/Symfony/Component/Debug/ErrorHandler.php [hunk 3]
```diff
@@ -112,7 +112,7 @@ class ErrorHandler
      *
      * @return self The registered error handler
      */
-    public static function register(self $handler = null, $replace = true): self
+    public static function register(self $handler = null, bool $replace = true): self
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Debug/ErrorHandler.php]: now that this class has been deprecated, all changes should be reverted (and the class should be removed actually)
see other classes too (should be easy to spot by rebasing)

Comment (Matts) [reply src/Symfony/Component/Debug/ErrorHandler.php]: Reverted all changes in class. With deprecated you mean final? I couldn't find any others in the namespace.

File: src/Symfony/Component/Debug/DebugClassLoader.php (hunk 2 above)
Comment (Matts) [reply src/Symfony/Component/Debug/DebugClassLoader.php]: reverted annotation and removed default null.

File: src/Symfony/Component/Debug/ErrorHandler.php (hunk 3 above)
Comment (stof) [reply src/Symfony/Component/Debug/ErrorHandler.php]: no, with deprecated, we mean deprecated.

File: src/Symfony/Component/Debug/DebugClassLoader.php [hunk 4]
```diff
@@ -129,21 +129,19 @@ public static function disable()
     }
 
     /**
-     * @return string|null
+     * @param string $class
```
Comment (stof) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: this is wrong. You must not remove the the doc for the return type (and adding the doc for the param is useless now, as it duplicates the typehint)

File: src/Symfony/Component/Debug/ExceptionHandler.php [hunk 5]
```diff
@@ -60,11 +60,11 @@ public function __construct(bool $debug = true, string $charset = null, $fileLin
      *
      * @param bool        $debug          Enable/disable debug mode, where the stack trace is displayed
      * @param string|null $charset        The charset used by exception messages
-     * @param string|null $fileLinkFormat The IDE link template
+     * @param string|FileLinkFormatter|null $fileLinkFormat The IDE link template
```
Comment (stof) [inline src/Symfony/Component/Debug/ExceptionHandler.php]: this should be added in older branches

**Comment by nicolas-grekas** [comment]
This PR is going to be obsoleted by https://github.com/symfony/symfony/pull/32471

**Comment by nicolas-grekas** [comment]
The Debug component is now gone, long live ErrorHandler. New PR welcome I suppose (please have a look at comments here of course)

**Pull request closed without merging**